pkg/controller/cloudlet_api_test.go

```diff
@@ -1125,3 +1127,114 @@ func testCloudletEdgeboxOnly(t *testing.T, ctx context.Context, cloudlet edgepro
+	require.Nil(t, err)
+	info := testutil.CloudletInfoData()[2]
+	info.Key = cloudlet.Key
+	info.State = dme.CloudletState_CLOUDLET_STATE_READY
+	apis.cloudletInfoApi.Update(ctx, &info, 0)
+	err = waitForState(&cloudlet.Key, edgeproto.TrackedState_READY, apis)
+	require.Nil(t, err, "cloudlet obj created")
+	// check that static fqdn and running match
+	cloudletObj := edgeproto.Cloudlet{}
+	ok := apis.cloudletApi.store.Get(ctx, &cloudlet.Key, &cloudletObj)
+	require.True(t, ok)
+	require.Equal(t, cloudletObj.RootLbFqdn, cloudletObj.StaticRootLbFqdn)
+
+	// Add some clusterInsts, appInsts
+	clusterInst := testutil.ClusterInstData()[5]
+	clusterInst.CloudletKey = cloudlet.Key
+	err = apis.clusterInstApi.CreateClusterInst(&clusterInst, testutil.NewCudStreamoutClusterInst(ctx))
+	require.Nil(t, err, "Create ClusterInst")
+	clusterObj := edgeproto.ClusterInst{}
+	ok = apis.clusterInstApi.store.Get(ctx, &clusterInst.Key, &clusterObj)
+	require.True(t, ok)
+	originalFqdn := clusterObj.Fqdn
+	// Set up a different appDNSRoot
+	*appDNSRoot = "new.and.improved.dns.com"
+	// Cloudlet has to be in maintenance mode
+	err = apis.cloudletApi.ChangeCloudletDNS(&cloudlet.Key, testutil.NewCudStreamoutCloudlet(ctx))
+	require.NotNil(t, err)
+	require.Contains(t, err.Error(), "maintenance mode")
+
+	// Update cloudlet maintenance mode
+	// TODO - 1160 - 1203 is duplicate code, should fix this
+	stateTransitions := map[dme.MaintenanceState]dme.MaintenanceState{
+		dme.MaintenanceState_FAILOVER_REQUESTED:    dme.MaintenanceState_FAILOVER_DONE,
+		dme.MaintenanceState_CRM_REQUESTED:         dme.MaintenanceState_CRM_UNDER_MAINTENANCE,
+		dme.MaintenanceState_NORMAL_OPERATION_INIT: dme.MaintenanceState_NORMAL_OPERATION,
+	}
+
+	cancel := apis.cloudletApi.cache.WatchKey(&cloudlet.Key, func(ctx context.Context) {
+		cl := edgeproto.Cloudlet{}
+		if !apis.cloudletApi.cache.Get(&cloudlet.Key, &cl) {
+			return
+		}
+		switch cl.MaintenanceState {
+		case dme.MaintenanceState_FAILOVER_REQUESTED:
+			info := edgeproto.AutoProvInfo{}
+			if !apis.autoProvInfoApi.cache.Get(&cloudlet.Key, &info) {
+				info.Key = cloudlet.Key
+			}
+			info.MaintenanceState = stateTransitions[cl.MaintenanceState]
+			apis.autoProvInfoApi.cache.Update(ctx, &info, 0)
+		case dme.MaintenanceState_CRM_REQUESTED:
+			fallthrough
+		case dme.MaintenanceState_NORMAL_OPERATION_INIT:
+			info := edgeproto.CloudletInfo{}
+			if !apis.cloudletInfoApi.cache.Get(&cloudlet.Key, &info) {
+				info.Key = cloudlet.Key
+			}
+			info.MaintenanceState = stateTransitions[cl.MaintenanceState]
+			apis.cloudletInfoApi.cache.Update(ctx, &info, 0)
+		}
+	})
+
+	defer cancel()
+
+	cloudlet.MaintenanceState = dme.MaintenanceState_MAINTENANCE_START
+	cloudlet.Fields = append(cloudlet.Fields, edgeproto.CloudletFieldMaintenanceState)
+	err = apis.cloudletApi.UpdateCloudlet(&cloudlet, testutil.NewCudStreamoutCloudlet(ctx))
+	require.Nil(t, err, fmt.Sprintf("update cloudlet maintenance state"))
```

Comment:
I couldn't use `cloudletApi.store.Put()` directly, since it does object validation and wants Maintenance_start instead of Under_maintenance, but I was able to just use `cloudetApi.setMaintenanceState()` here.